Traffic Boost: Add backend implementation

vaurdan opened the pull request:
## Description

This PR implements the backend for Traffic Boost, that integrates with the Suggestions API to generate traffic boost suggestions.

Alongside the new services added - `suggest-inbound-links` and `suggest-inbound-link-positions`, a new REST API endpoint (`traffic-boost`) has been added that allows manipulating the Traffic Boost data (add, remove and list suggestions and existing inbound links).

This PR adds a new composer dependency - `masterminds/HTML5` - that helps with the DOM manipulation of the posts where links are added and removed. 

## Motivation and context
 * Implement the Traffic Boost backend as part of #2887
 * Closes #2887 as the MVP of Traffic Boost is now completed. New improvements will be added in different PRs.

## How has this been tested?
Tested locally, on a clean environment and on a clone environment of an existing site. This has also been tested on a staging environment of a VIP site. 

No new tests have been implemented yet, but existing tests do pass.

## Screenshots (if appropriate)
![image](https://github.com/user-attachments/assets/576a1d15-fa18-4cd0-820e-8731bbfa6436)
![image](https://github.com/user-attachments/assets/99ed9cd6-9090-4b2c-80ef-a2abcb90457b)


<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

- **New Features**
  - Introduced a new endpoint for managing traffic boost suggestions.
  - Enhanced capabilities for generating and managing inbound link suggestions, including improved handling of link placements and validations.
  - Added loading state management and error handling improvements across various components.

- **Refactor**
  - Upgraded routing and API integrations for more robust error management and smoother interactions.
  - Streamlined logic across preview, dashboard, and suggestion components.

- **Style**
  - Refined interface styling for previews, loading indicators, and interactive elements.

- **Tests**
  - Expanded test coverage to validate the new traffic boost endpoint integration.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

vaurdan committed:
Initial implementation of the suggest-inbound-links endpoint

vaurdan committed:
Add generate button functionality

vaurdan committed:
Merge branch 'refs/heads/add/traffic-boost' into add/traffic-boost-backend

vaurdan committed:
Implement generate/accept/discard suggestions and remove inbound link endpoints.

vaurdan committed:
Fix issue with title HTML encoding

vaurdan committed:
Merge branch 'add/traffic-boost' into add/traffic-boost-backend

# Conflicts:
#	package-lock.json
#	src/content-helper/common/settings/provider.tsx

vaurdan committed:
Fix linting and phpstan issues

vaurdan committed:
UI: Fix minor glitch with container height and long WP sidebars

vaurdan committed:
UI: Fix issue with merging links in the preview when highlight is disabled.

vaurdan committed:
UI: Fix issue with wp-admin sidebar collapsing

vaurdan committed:
Add accept and discard functionality

vaurdan committed:
Minor clean-up and improvements

vaurdan committed:
Add 'Remove Inbound' functionality

vaurdan committed:
Improve Smart Link placement and validation logic

vaurdan committed:
Implementation of the Update Inbound Link feature.

vaurdan committed:
Implementation of the Add Manual Link feature

vaurdan committed:
Implementation of the Add Manual Link feature

vaurdan committed:
Merge remote-tracking branch 'origin/add/traffic-boost-backend' into add/traffic-boost-backend

vaurdan committed:
Minor tweaks and "Regenerate suggestion" button.

vaurdan committed:
Improvements around existing link counting and highlighting.

vaurdan committed:
Improvements to the suggestion validation and handling

vaurdan committed:
Fix linting and phpstan

vaurdan committed:
Fix issue with unwanted JS loading in preview.

vaurdan committed:
Fix issue with outbound links highlight triggering and getting stuck on a loading state.

vaurdan committed:
Fix PHP linting

vaurdan committed:
Fix tests

vaurdan commented on src/Models/class-inbound-smart-link.php:
```diff
@@ -140,73 +257,724 @@
+		
+		// Query for any 'a' element that has a data-smartlink attribute matching our UID.
+		$query            = sprintf( '//a[@data-smartlink="%s"]', $this->uid );
+		$smart_link_nodes = $xpath->query( $query, $node );
```
Nice to see that SonarCloud now has a nice GitHub integration :)

acicovic commented on src/content-helper/dashboard-page/pages/traffic-boost/sidebar/components/post-details.tsx:
```diff
@@ -8,14 +8,16 @@ import { __ } from '@wordpress/i18n';
  */
 import { HydratedPost } from '../../../../../common/base-wordpress-provider';
 import { Thumbnail } from '../../../../../common/components/thumbnail';
+import { Spinner } from '@wordpress/components';
```
This should go into the WordPress dependencies section.

acicovic commented on src/content-helper/dashboard-page/pages/traffic-boost/sidebar/components/tabs/inbound-links-tab.tsx:
```diff
@@ -9,6 +9,7 @@ import { useDispatch, useSelect } from '@wordpress/data';
 import { TrafficBoostLink } from '../../../provider';
 import { TrafficBoostStore } from '../../../store';
 import { LinksList } from '../links-list/links-list';
+import { __ } from '@wordpress/i18n';
```
This should go into the WordPress dependencies section.

acicovic commented on src/services/class-base-service-endpoint.php:
```diff
@@ -99,6 +99,7 @@ protected function get_request_options( string $method ): array {
 		$options = array(
 			'headers' => $this->get_headers(),
 			'method'  => $method,
+			'timeout' => 60, // phpcs:ignore WordPressVIPMinimum.Performance.RemoteRequestTimeout.timeout_timeout
```
Should this be matching the timeout in `src/services/suggestions-api/endpoints/class-suggestions-api-base-endpoint.php` (90 seconds)?

vaurdan replied:
It's not necessary, since this higher timeout value is only required for Suggestions API calls, since it can take some time for the API to return a response.

acicovic commented on src/content-helper/editor-sidebar/smart-linking/class-smart-linking.php:
```diff
@@ -93,8 +93,8 @@ private function register_post_type(): void {
 				'hierarchical'    => false,
-				'public'          => false,
-				'show_ui'         => false,
+				'public'          => true, // TODO: revert to false.
+				'show_ui'         => true, // TODO: revert to false.
```
Are the todos in this file to be fixed in this PR?

acicovic commented on src/content-helper/editor-sidebar/smart-linking/utils.ts:
```diff
@@ -450,6 +450,7 @@ export function getAllSmartLinksInPost(): SmartLink[] {
 
 		const smartLink: SmartLink = {
 			uid,
+			smart_link_id: 0,
```
Code styling question: Do you have a convention for when to use camelCase or snake_case in these instances? `match` below uses camelCase.

vaurdan replied:
The pattern is basically when it's data that is coming from the REST API, it uses snake_case, when it's data managed by the JS UI, it's camelCase. So in this case, the `SmartLink` describes the smart link structure that is returned by the REST API, so it is `smart_link_id` instead of `smartLinkId`. 

Unfortunately, while this can be a bit confusing, it is enforced by the PHP linter.

acicovic replied:
Thank you!

acicovic commented on src/content-helper/dashboard-page/components/posts-table/component.tsx:
```diff
@@ -49,8 +53,9 @@ const PostInfo = ( { post }: { post: HydratedPost } ): React.JSX.Element => {
 					{ post.title.rendered !== ''
-						? post.title.rendered
-						: __( '(no title)', 'wp-parsely' ) }
+						? <div dangerouslySetInnerHTML={ { __html: post.title.rendered } } />
+						: __( '(no title)', 'wp-parsely' )
+					}
```
@vaurdan I also saw another use of `dangerouslySetInnerHTML` which wasn't flagged by CodeRabbit. Something we should check if these warnings have merit.

vaurdan replied:
@acicovic see[ my comment in the conversation above](https://github.com/Parsely/wp-parsely/pull/3149#discussion_r1983093841). It is safe in this context.

acicovic commented on src/content-helper/dashboard-page/components/posts-table/component.tsx:
```diff
@@ -182,6 +192,7 @@ export const PostsTable = ( {
 	className,
 	onPostClick,
 }: PostsTableType ): React.JSX.Element => {
+	// TODO: Add a global state to store the posts for faster loading.
```
Should this work be in another PR, or this one?

vaurdan replied:
It's optimization work for another PR 🙂

acicovic commented on src/services/suggestions-api/class-suggestions-api-service.php:
```diff
@@ -112,4 +115,38 @@ public function get_smart_links( string $content, $options = array(), array $url
+	/**
+	 * Gets suggested inbound links for the given URL.
+	 *
+	 * @since 3.17.0
```
Is this `@since` version correct?

acicovic commented on src/services/suggestions-api/endpoints/class-endpoint-suggest-inbound-link-positions.php:
```diff
@@ -0,0 +1,138 @@
+	}
+
+	/**
+	 * Gets suggested inbound link position for a given source and destination post using the Parse.ly
```
Should `position` be `positions` here?

acicovic commented on src/services/suggestions-api/endpoints/class-endpoint-suggest-inbound-link-positions.php:
```diff
@@ -0,0 +1,138 @@
+	/**
+	 * Executes the API request.
+	 *
+	 * @since 3.17.0
```
This `@since` version seems incorrect.

acicovic commented on src/services/suggestions-api/endpoints/class-endpoint-suggest-inbound-links.php:
```diff
@@ -0,0 +1,153 @@
+		$request_body = array(
+			'canonical_url' => $post_url,
+			'output_config' => array(
+				'blending_weight' => 0.9,
```
We probably need to change this to `performance_blending_weight`.

acicovic commented on src/services/suggestions-api/endpoints/class-endpoint-suggest-inbound-links.php:
```diff
@@ -0,0 +1,153 @@
+	/**
+	 * Executes the API request.
+	 *
+	 * @since 3.17.0
```
This `@since` version seems incorrect.

acicovic commented on src/content-helper/dashboard-page/pages/traffic-boost/single-post-component.tsx:
```diff
@@ -49,50 +55,114 @@ export const TrafficBoostPostPage = (): React.JSX.Element => {
 		}
-	}, [] );
+
+		// Grabs the sidebar width so it can be used to calculate the container width.
```
Does this comment need to be placed somewhere else?

acicovic commented on src/content-helper/dashboard-page/pages/traffic-boost/single-post-component.tsx:
```diff
@@ -128,90 +198,223 @@ export const TrafficBoostPostPage = (): React.JSX.Element => {
+	 */
+	useEffect( () => {
+		if ( error ) {
+			// TODO: better error handling.
```
Is this to be implemented within this PR?

acicovic commented on src/content-helper/dashboard-page/pages/traffic-boost/single-post-component.tsx:
```diff
@@ -128,90 +198,223 @@ export const TrafficBoostPostPage = (): React.JSX.Element => {
+				if ( err instanceof ContentHelperError ) {
+					setError( err );
+				} else {
+					setError( new ContentHelperError( 'Failed to fetch inbound links', ContentHelperErrorCode.FetchError ) );
```
Does the string need to be translatable?

acicovic commented on src/content-helper/dashboard-page/pages/traffic-boost/single-post-component.tsx:
```diff
@@ -128,90 +198,223 @@ export const TrafficBoostPostPage = (): React.JSX.Element => {
+				const fetchedSuggestions = await trafficBoostProvider.getExistingSuggestions( parseInt( postId ) );
+
+				// If there are no suggestions, trigger the generation of suggestions.
+				if ( fetchedSuggestions.length === 0 ) {
```
Let's use a Yoda condition here, as done in the rest of the file.

acicovic commented on src/rest-api/content-helper/class-endpoint-traffic-boost.php:
```diff
@@ -0,0 +1,770 @@
+ * Parse.ly Content Helper `/traffic-boost` API endpoint class
+ *
+ * @package Parsely
+ * @since   3.17.0
```
Lots of `@since 3.17.0` in this file.

acicovic commented on src/rest-api/content-helper/class-endpoint-traffic-boost.php:
```diff
@@ -0,0 +1,770 @@
+			'/generate',
+			array( 'POST' ),
+			array( $this, 'generate_link_suggestions' ),
+			array(
```
We aren't setting `required` here, in contrast to other route registrations.

vaurdan replied:
Right, none of these parameters are required because they have default values.

acicovic replied:
I guess my confusion/question comes from the fact that we're explicitly setting `required` to `true` or `false` in other routes below. Not that it's important if things work as they should. Fee free to ignore this, I'm just setting it to unresolved to make it more easily visible.

vaurdan replied:
Ah gotcha! I'm going to set require to false explicitly so it is consistent and easier to understand when reading the file.

acicovic commented on src/content-helper/dashboard-page/pages/traffic-boost/provider.ts:
```diff
@@ -78,298 +145,377 @@ export class TrafficBoostProvider extends BaseWordPressProvider {
+		}
+
+		throw new ContentHelperError(
+			response.message ?? 'Unknown error.',
```
Do we need to make this string translatable?

acicovic commented on src/Models/class-smart-link.php:
```diff
@@ -37,6 +37,15 @@ class Smart_Link extends Base_Model {
+	/**
+	 * The source post object.
+	 *
+	 * @since 3.16.0
```
Should this be 3.18.0?

acicovic commented on src/Models/class-smart-link.php:
```diff
@@ -318,12 +366,21 @@ public function save(): bool {
+	/**
+	 * Updates the UID of the smart link.
+	 *
+	 * @since 3.16.0
```
3.18?

acicovic commented on src/Models/class-smart-link.php:
```diff
@@ -399,8 +471,42 @@ private function get_string_meta( string $meta_key, string $default_value = '' )
+	/**
+	 * Gets a boolean meta value from the smart link post.
+	 *
+	 * @since 3.16.0
```
3.18?

acicovic commented on src/Models/class-smart-link.php:
```diff
@@ -624,45 +852,167 @@ public static function get_outbound_smart_links( int $post_id ): array {
+	/**
+	 * Gets the cache group for a single smart link.
+	 *
+	 * @since 3.16.0
```
I see v3.16.0 here and all functions below, which probably means to be 3.18.0.

acicovic commented on src/Models/class-smart-link.php:
```diff
@@ -624,45 +852,167 @@ public static function get_outbound_smart_links( int $post_id ): array {
+	}
+
+	/**
+	 * Flushes the cache for all smart links in a post.
```
This comment seems incorrect.

vaurdan replied:
It's actually correct, it does flushes the caches of all the smart links that are associated with the provided post ID. I have tweaked the comment to make it more clear, though.

acicovic replied:
Weird, it's indeed correct. Maybe I misread something.

acicovic commented on src/Models/class-inbound-smart-link.php:
```diff
@@ -72,11 +96,83 @@ public function to_array(): array {
+
+		if ( ! $post instanceof WP_Post ) {
+			if ( $wp_error ) {
+				return new \WP_Error( 'traffic_boost_invalid_post', 'Invalid post' );
```
Many `WP_Error`s in this file are with untranslatable strings.

acicovic commented on src/Models/class-inbound-smart-link.php:
```diff
@@ -231,6 +999,259 @@ public static function from_smart_link( Smart_Link $smart_link ): Inbound_Smart_
+		// Get all inbound smart links for the post.
+		$smart_links = self::get_inbound_smart_links( $post_id, Smart_Link_Status::PENDING );
+
+		// Filter out the ones that are already applied.
```
Can we get Smart Links with an `applied` status when we've specified `pending` above?

vaurdan replied:
Not in this case, but that's intended: `get_existing_suggestions` should only return suggestions, i.e. smart links that have not been applied - pending. 

The filtering below is just a double check to make sure that if for some reason the function returns an applied smart link, it would be filtered out. 

Do you think we should remove it?

acicovic replied:
My thoughts:

If it's possible to get non-desired results from the function, maybe the best option is to have any safeguards in the function itself if we can. Applying external safeguards means that we'd need to duplicate them whenever using the function, and we might forget to do that.

Alternatively, if this is needed, I'd update the comment to say that it's a safeguard against a potential situation, as the current phrasing might be interpreted that getting back applied links is kind of expected.

vaurdan replied:
I agree with you! I'm going to remove the `array_filter` and save some CPU cicles 🙂  The database select in `get_inbound_smart_links` should be enough to make sure we're only getting pending smart links.

acicovic committed:
Fix some nitpicks (whitespace, import placements, text case)

acicovic commented:
Hey @vaurdan, thanks for working on this, I've completed my review. I also wanted to note that I'm getting these PHPStan errors:

```
 ------ ---------------------------------------------------------------------- 
  Line   src/Models/class-inbound-smart-link.php                               
 ------ ---------------------------------------------------------------------- 
  501    Parameter #1 $node of method                                          
         Parsely\Models\Inbound_Smart_Link::find_text_node() expects DOMNode,  
         mixed given.                                                          
 ------ ---------------------------------------------------------------------- 

 ------ -------------------------------------------------------------------- 
  Line   src/UI/class-dashboard-page.php                                     
 ------ -------------------------------------------------------------------- 
  71     Strict comparison using === between 0 and 1|2|false will always     
         evaluate to false.                                                  
         💡 Because the type is coming from a PHPDoc, you can turn off this   
            check by setting treatPhpDocTypesAsCertain: false in your        
            phpstan.neon.                                                    
  131    Strict comparison using === between 0 and 1|2|false will always     
         evaluate to false.                                                  
         💡 Because the type is coming from a PHPDoc, you can turn off this   
            check by setting treatPhpDocTypesAsCertain: false in your        
            phpstan.neon.                                                    
 ------ -------------------------------------------------------------------- 

 [ERROR] Found 3 errors                                                         
```

vaurdan commented:
@acicovic I have addressed your feedback! I will now move to check Code Rabbit's feedback and fix any issues in another commit. 

>  I also wanted to note that I'm getting these PHPStan errors:

Interesting, I don't see the `src/Models/class-inbound-smart-link.php` error in my PHPStan. Running with phpstan 1.12.20, after running `composer update`.

```bash
$ composer static-analysis
> @php ./vendor/bin/phpstan analyze --memory-limit=2G
Note: Using configuration file /Users/henrique/Repositories/wp-parsely/phpstan.neon.
 165/165 [▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓] 100%

 ------ -------------------------------------------------------------------------------------------------------------------------------------------- 
  Line   src/UI/class-dashboard-page.php                                                                                                             
 ------ -------------------------------------------------------------------------------------------------------------------------------------------- 
  :71    Strict comparison using === between 0 and 1|2|false will always evaluate to false.                                                          
         💡 Because the type is coming from a PHPDoc, you can turn off this check by setting treatPhpDocTypesAsCertain: false in your phpstan.neon.  
  :131   Strict comparison using === between 0 and 1|2|false will always evaluate to false.                                                          
         💡 Because the type is coming from a PHPDoc, you can turn off this check by setting treatPhpDocTypesAsCertain: false in your phpstan.neon.  
 ------ -------------------------------------------------------------------------------------------------------------------------------------------- 
                                                    
 [ERROR] Found 2 errors                                                                                                 
                                                                                                                        
Script @php ./vendor/bin/phpstan analyze --memory-limit=2G handling the static-analysis event returned with error code 1                                                                                                  
                                                                                                                        
$ php ./vendor/bin/phpstan --version
PHPStan - PHP Static Analysis Tool 1.12.20
```

I have fixed these issues though. 

@coderabbitai pause

vaurdan committed:
Address @acicovic feedback

vaurdan committed:
Address @acicovic feedback

vaurdan commented on src/content-helper/dashboard-page/pages/traffic-boost/sidebar/components/links-list/single-link.tsx:
```diff
@@ -73,9 +73,10 @@ export const SingleLink = ( {
-				<div className="single-link-title">
-					{ suggestedPost.title.rendered }
-				</div>
+				<div
+					className="single-link-title"
+					dangerouslySetInnerHTML={ { __html: suggestedPost.title.rendered } }
+				/>
```
This is safe, since we're using the `wpPost.title.rendered` propriety, which is intended to be outputted as HTML. This usage of `dangerouslySetInnerHTML` is required to properly render the HTML entities that might be present in the title.

vaurdan commented on src/content-helper/dashboard-page/pages/traffic-boost/preview/components/preview-header.tsx:
```diff
@@ -145,18 +149,26 @@ export const PreviewHeader = ( {
 			<div className="traffic-boost-preview-info">
 				<div className="traffic-boost-preview-info-title">
-					{ activeLink?.targetPost?.title.rendered }
+					<div dangerouslySetInnerHTML={ { __html: activeLink?.targetPost?.title.rendered } } />
```
Same as above.

vaurdan commented on src/content-helper/dashboard-page/pages/traffic-boost/provider.ts:
```diff
@@ -78,298 +145,377 @@ export class TrafficBoostProvider extends BaseWordPressProvider {
 	 *
-	 * @return {Promise<TrafficBoostLink[]>} The list of boost links.
+	 * @return {Promise<TrafficBoostLink>} The suggestion.
 	 */
-	public async getInboundLinks( postId: number ): Promise<TrafficBoostLink[]> {
-		// Request inbound smart links for the post.
-		const inboundSmartLinks = await this.getInboundSmartLinks( postId );
-
-		if ( inboundSmartLinks.length === 0 ) {
-			return [];
-		}
+	public createSuggestion( post: HydratedPost ): TrafficBoostLink {
+		return {
+			uid: `suggestion-${ post.id }-${ Date.now() }`,
+			targetPost: post,
+			isSuggestion: true,
+		};
+	}
```
I believe this has been fixed on recent commits. Can you check?

vaurdan replied:
Ah nevermind. It's on the docblock. Fixed, thank you.

vaurdan committed:
Apply Code Rabbit's feedback

vaurdan committed:
Fix SonarCloud issue on regex groups

vaurdan committed:
Resolve SonarCloud issue with operator precedence

vaurdan commented:
@acicovic SonarCloud is complaining about duplicate lines, but it's actually a false positive and I don't think it warrants a fix. So feel free to ignore the failing status of it, I have addressed all the other issues flagged by it.

acicovic commented:
Hey @vaurdan, I'm completing my review and I've responded to a couple of conversations. I haven't re-checked PHPStan, but if this is an issue it will surely come up again when work is done against the branch.

vaurdan committed:
Implement @acicovic suggestions

vaurdan commented:
@acicovic I have addressed your comments in the latest commit, feel free to take a look!

acicovic reviewed (approved):
Thank you for working on this! 🙂
